core/src/main/scala/com/rallyhealth/vapors/factfilter/data/FactTypeSet.scala

```diff
@@ -2,34 +2,32 @@ package com.rallyhealth.vapors.factfilter.data
 
 import cats.data.{NonEmptyList, NonEmptyMap, NonEmptySet}
 
-import scala.collection.immutable
+import scala.collection.immutable.{SortedMap, SortedSet}
 
-final case class FactTypeSet[A] private (typeMap: NonEmptyMap[String, FactType[A]]) {
+final case class FactTypeSet[A] private (typeMap: NonEmptyMap[String, FactType[A]]) extends AnyVal {
 
-  lazy val typeList: NonEmptyList[FactType[A]] = typeMap.toNel.map(_._2)
+  def typeList: NonEmptyList[FactType[A]] = NonEmptyList.fromListUnsafe(typeMap.toSortedMap.values.toList)
 
-  lazy val typeSet: NonEmptySet[FactType[A]] = typeList.toNes
+  def typeSet: NonEmptySet[FactType[A]] = NonEmptySet.fromSetUnsafe(SortedSet.from(typeMap.toSortedMap.values))
 
-  private val castPF: PartialFunction[Fact, TypedFact[A]] = {
-    case fact @ TypedFact(factType, _) if typeMap(factType.fullName).contains(factType) =>
-      // Justification: This checks equality of the FactType at runtime after safely casting the value
-      fact.asInstanceOf[TypedFact[A]]
-  }
-
-  private val castF: Fact => Option[TypedFact[A]] = castPF.lift
+  def unapply(fact: Fact): Option[TypedFact[A]] = collector.lift(fact)
 
   /**
     * Checks the [[FactType]] for equality with one of the types in this set.
     *
     * @return the [[TypedFact]] as the expected type.
     */
-  def cast(fact: Fact): Option[TypedFact[A]] = castF(fact)
+  def cast(fact: Fact): Option[TypedFact[A]] = collector.lift(fact)
 
   /**
     *
     * @return a partial function for collecting facts of a specific type
     */
-  def collector: PartialFunction[Fact, TypedFact[A]] = castPF
+  def collector: PartialFunction[Fact, TypedFact[A]] = {
```

Comment:
There are... several issues with this function. Maybe there should be a requirement that there are enums of Facts provided?

Does this work by type inference? I guess by the type parameter of the class? Is that reliable?

Reply:
I am thinking of maybe using Shapeless's `TypeCase` instead of TypeTag, but it doesn't seem necessary at this point (maybe for Scala 3 cross-compatibility)